### DISCUSSION TITLE
Do NMEA2000 spec define range for "fast-packet" PGNs?

### DISCUSSION THREAD

**Discussion by aldas** [canboat/canboat#375] [Q&A]
SAE J1939 has specific ranges of PGNs for addressed and broadcast packets.  See https://copperhilltech.com/blog/design-of-proprietary-parameter-group-numbers-pgns/

Do NMEA2000 have something similar in regards of "fast-packets"? It would seem logical in some sense to have defined range as then it would be very easy for micro-controllers to know when multi packet assembly is required (allthough microcontroller probably filter only PGN they support). 
Looking at Canboat pgns.h it seems that there are some ranges but then there are "mixed" ranges

https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L1672-L1672
https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L1873

and there is this range "single/fast"

https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L2217

It seems little bit weird that they would have this mixed range as it makes "life harder" for to implement NMEA2000 protocol and compared to NMEA0183 seems that there have been a lot of more thought put into protocol design but yet there is "mixed".

Maybe someone knows or has friend that know NMEA2000 spec secrets? Is it possible to deduce 'fast-packet' just from PGN number without having PGN database?

----
p.s. actual fast/non-fast can listed as such:
1) this command will list all non-fast packet PGNS `jq '.PGNs[] | select(.Type != "Fast") | .PGN' docs/canboat.json | sort | uniq`
2) this will list all fast-packet PGNS `jq '.PGNs[] | select(.Type == "Fast") | .PGN' docs/canboat.json | sort | uniq`

**Comment by aldas** [comment]
I would assume there is no single rule when it comes to proprietary PGNs. For example 130824 has [single](https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L5828) and [fast-packet](https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L5840) variants. 

I did Github code search and here are some remarks/approaches to this problem:
* [ttlappalainen/NMEA2000](https://github.com/ttlappalainen/NMEA2000/blob/65a7c605682ea14324380f87e5c4b3c494cdc1b9/src/NMEA2000.cpp#L1502) has some helper functions with list of PGNs to deduce if packet is fast. Also supports registering your own PGN as fast.
* [erkenny/simIntegration](https://github.com/erkenny/simIntegration/blob/3a5643d735163b430087c35a01b0a1ff6dd116c6/DsJ1939Def.h#L37) says that you can not deduce that from PGN.


I am wondering how for example Actisense device, that will output fully assembled packets, knows to handle packets like 130824

**Comment by aldas** [comment]
@keesverruijt  PGN 130824 "B&G: Wind data" is marked as "Single"  but most of the samples seem to be more than 8 bytes.  Could it be that "B&G: Wind data" should be actually fast-packet?



-------


Note: `B&G` manufacturer id on 381 which means that if packet starts with `7d` it is `B&G`

`grep 130824 samples/*`


```
samples/ws320.raw:2018-02-14T15:55:12.954Z,3,130824,3,255,8,7d,99,f8,40,86,72,80,05
samples/ws320.raw:2018-02-14T15:55:13.080Z,3,130824,3,255,8,7d,99,f8,40,fe,72,80,05
```

```
samples/susteranna-actisense-serial.raw:2016-04-09T16:41:39.970Z,7,130824,17,255,20,7d,99,3c,40,77,9f,fe,ff,9e,40,92,f0,ff,ff,44,40,f7,c0,00,00
samples/susteranna-actisense-serial.raw:2016-04-09T16:41:40.044Z,7,130824,17,255,20,7d,99,3c,40,99,af,fe,ff,9e,40,d1,1b,00,00,44,40,fb,3f,00,00
```

There are occasional shorter packages
```
samples/susteranna-actisense-serial.raw:2016-04-09T16:41:35.913Z,3,130824,21,255,2,7d,99
```

This `susteranna2020.raw` seems to be series of not yet assembled frames for 130824 fast-packet
```
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a0,2e,7d,99,37,21,fa,3f
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a1,38,21,34,04,7e,20,a9
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a2,01,7c,20,00,00,32,21
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a3,c0,e5,33,21,c5,22,3a
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a4,21,00,00,1d,21,00,00
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a5,50,21,61,96,51,21,ce
samples/susteranna2020.raw:2020-08-22T13:52:36.950Z,3,130824,27,255,8,a6,08,52,21,32,f7,ff,ff
```

--------
p.s. 130824 "Maretron: Annunciator" samples are in
```
samples/dirona-actisense-serial.raw:2016-02-28T19:57:54.001Z,2,130824,24,255,9,89,98,00,00,ff,ff,ff,ff,ff
samples/dirona-actisense-serial.raw:2016-02-28T19:58:03.921Z,2,130824,24,255,9,89,98,00,00,ff,ff,ff,ff,ff
```
```
samples/merrimac-2022.raw:2022-09-10T12:06:59.345Z,3,130824,26,255,84,7d,99,64,20,53,00,0d,21,67,00,0e,41,71,a1,b2,1f,0f,41,93,51,3c,03,69,20,53,7a,d3,20,38,21,81,40,70,d3,02,00,9a,20,4a,88,82,20,00,00,0a,21,98,02,0c,21,b8,02,75,40,20,6c,fb,ff,09,41,ff,ff,ff,ff,0b,41,ff,ff,ff,ff,d0,40,30,0e,a4,0d,7f,20,00,00,9d,20,fe,98
samples/merrimac-2022.raw:2022-09-10T12:06:59.487Z,3,130824,26,255,14,7d,99,50,21,cb,08,51,21,d1,fd,52,21,2f,02
```

**Comment by keesverruijt** [comment]
> I would assume there is no single rule when it comes to proprietary PGNs. For example 130824 has [single](https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L5828) and [fast-packet](https://github.com/canboat/canboat/blob/53a4f3548546e9f68f0c626d0e5ef38341dac97a/analyzer/pgn.h#L5840) variants.

Yeah, I think that is actually an incorrect flag for the B&G wind data.

My current "best guess" (which I really believe to be correct) is the following table originating from `canboat.xsl`:

https://canboat.github.io/canboat/canboat.xml#pgn-ranges



Range Hex | Range Dec | PDU | Step | Number of possible PGNs | Use | Framing
-- | -- | -- | -- | -- | -- | --
0xE800-0xEE00 | 59392 - 60928 | PDU1 | 256 | 7 | ISO 11783 (protocol) | Single frame
0xEF00 | 61184 | PDU1 |   | 1 | Manufacturer proprietary | Single frame
0xF000-0xFEFF | 61440 - 65279 | PDU2 | 1 | 3840 | Standardized | Single frame
0xFF00-0xFFFF | 65280 - 65535 | PDU2 | 1 | 256 | Manufacturer proprietary | Single frame
0x1ED00-0x1EE00 | 126208 - 126464 | PDU1 | 256 | 2 | Standardized (protocol) | Single frame
0x1EF00 | 126720 | PDU1 |   | 1 | Manufacturer proprietary | Fast packet
0x1F000-0x1FEFF | 126976 - 130815 | PDU2 | 1 | 3840 | Standardized | Mixed single/fast
0x1FF00-0x1FFFF | 130816 - 131071 | PDU2 | 1 | 256 | Manufacturer proprietary | Fast packet

As you can see, there is _one_ range which contains both single and fast packets, unfortunately. But none of them are manufacturer proprietary. If you think about it, if they did this then no manufacturer mixing would be possible.

I created an issue #377 to make sure such bugs cannot reappear, and update the B&G Wind Data 130824 definition.

**Comment by keesverruijt** [comment]
Also, you _CANNOT_ rely on looking at canboat format raw logs to see if the packet was sent 'single' or 'fast' framed, as you don't generally know how the file was generated; most of them stem from `actisense-serial` or `ikonvert-serial` that already do fast packet assembly.

**Comment by keesverruijt** [comment]
PS once I started writing down the PGN ranges in hex, and add PDU1/PDU2 labels, the ranges suddenly made a lot more sense...

@ttlappalainen, would you care to share your thoughts?

**Comment by ttlappalainen** [comment]
I have not found any logic and think you just need to have them listed. E.g., 130306 is single frame while 130074 is fastpacket and they are PDU2 with only global addess.

**Comment by Elmue** [comment]
I did a reverse engineering of an NMEA software (I will not mention which one here) and I found an interesting logic that they use to distinguish between single packets and fast packets:

For all known PGNs (the PGNs which the software can decode) they have a lookup table where for each PGN is defined separately if it uses single or fast packets.

But interesting is the fallback mechanism which they use for unknown PGNs:

**Only** for the following ranges they display that it is an "Unknown Fast Packet PGN":
```
0x1F000 until 0x1F01A
0x1F100 until 0x1F11A
0x1F200 until 0x1F21A
0x1F300 until 0x1F31A
-----------------------
0x1F500 until 0x1F51A
0x1F600 until 0x1F61A
-----------------------
0x1F800 until 0x1F81A
0x1F900 until 0x1F91A
0x1FA00 until 0x1FA1A
0x1FB00 until 0x1FB1A
0x1FC00 until 0x1FC1A
0x1FD00 until 0x1FD1A
0x1FE00 until 0x1FE1A
-----------------------
0x1FF00 until 0x1FFFF  (entire range)
```

**ATTENTION:**
That these are **reserved** for fast packets does not mean that all messages in this range **must** be fast packets.
They may also use single packets in these ranges,

On the other hand for the following ranges they display that it is an "Unknown Single Packet PGN"
```
0x1F400 until 0x1F4FF
0x1F700 until 0x1F7FF
```

After finding this algorithm I compared it with my list of fast packet PGNs that I use currently and very interestingly not even one  fast packet PGN exists where the low byte is greater than 0x1A (except in the last range 1FFXX).
And also not even one fast packet PGN exists in the range 0x1F4XX or 0x1F7XX.
This confirms that this rule in deed exists.
So it seems that the NMEA has made this secret agreement with the manufacturers.
                
I find it interesting that there are only very few PGNs reserved for fast packets.
It seems that the NMEA wants to encourage the manufacturers to use single packets wherever possible, which makes much sense to reduce bus traffic. 

Several manufacturers seem to have no conscience that CAN bus has been designed to send small packets.
The design of String_LZ always wastes one useless byte: You need either a length byte or a zero termination, but never both.
And a message like in logfile pgn129808.raw where they send a String_LAU like this:
12,01,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF
(length = 18 bytes full of FF) shows a firmware developer without any conscience of reducing useless bus traffic.

@Kees:
May be you add this to chapter "PGN ranges" on your page canboat.html
